File: problem_10.js
// The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.

// Find the sum of all the primes below two million.


# def check_p(num)
#   primes_hash = {}
#   (2..(Math.sqrt(num))).each do |n|
#     if num % n == 0 && primes_hash[n] == nil
#       primes_hash[n] ||= 0
#       primes_hash[n] += 1
#     end
#   end
# end
# "require 'pry'
def summation_of_primes(max)
  array_of_nums = (0...max).to_a
  array_of_nums[0] = 0
  array_of_nums[1] = 0
  #skip index 1 as thats number 1 and don't want to add that in later, 0 by default wont be added in
  prod = 0
  array_of_nums.each do |num|
    prod = 0
    next if num == 0
    while prod ** 2 < max
      (2..(max)).each do |mult|
        prod = num * mult
        array_of_nums[prod] = 0
      end
    end
  end

  p array_of_nums.compact.inject(&:+)
end

def seive

end

(2...2000000).to_a.map {|num| true}

summation_of_primes(200)


# require 'prime'
# puts Prime.each(1999999).inject(&:+)

File: problem_10.rb
# The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.

# Find the sum of all the primes below two million.
# Input: an integer n > 1
 
# Let A be an array of Boolean values, indexed by integers 2 to n,
# initially all set to true.
 
#  for i = 2, 3, 4, ..., not exceeding √n:
#   if A[i] is true:
#     for j = i2, i2+i, i2+2i, ..., not exceeding n :
#       A[j] := false


# cheap solution
# require 'prime'
# puts Prime.each(1999999).inject(&:+)

# Sieve of E solution
class JamesSieve
  def initialize(top)
    @top = top
    @topsquared = (top ** (0.5)).floor
    @array_of_nums=(2..top).to_a
  end


  def cross_off
    (0..(@array_of_nums.index(@topsquared))).each do |i|
      n = @array_of_nums[i]
      (i + n).step((@top-1),n) {|mult| @array_of_nums[mult] = nil} unless n.nil?
    end
    p @array_of_nums.compact.inject(&:+)
  end
end

puts JamesSieve.new(2000000).cross_off
# 142913828922


